Ensure cart changes remain after using back button to get back to the cart

opr opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
### Description
If the customer alters their cart, i.e. changes the quantity or adds an item, then navigates away from the page, for example by clicking the "Proceed to checkout" button, and then returns to the cart via the back button in their browser, then their changes will not be shown. The client and the server are out of sync.

This happens because we print the cart data onto the page at page load, and the cart hydrates itself based on this. When the user returns to the page using the back button, the data that were printed onto the page are stale, so the cart hydrates itself with the data that were present when the page was first loaded.

### Changes made

#### Source Code
- Added a new action to `assets/js/data/cart/actions` called: `cartDataIsStale`. This will set a flag in the store indicating the cart data is out of sync with the server. This is based on the check detailed below.
- Add `cartUpdateMiddleware` to `assets/js/middleware/cart-update.ts` which will write the timestamp of the last cart update to the browser's local storage.
- Update the `CartSchema` to add a new field: `generated_timestamp` this will allow us to know when the data were generated so we can compare it to the last time the cart was updated (which is stored in local storage).
- Add a check to the `withStoreCartApiHydration` hook that will compare the `generated_timestamp` time to the one in local storage to determine if the cart needs to be fetched from the API. 
- The cart data will not be rehydrated from `wcSettings` if it is determined that the data is stale.

#### Test code
This PR also introduces some new e2e tests to test this functionality. I needed to create some utility functions to enable testing the front end. The details of this follow.

- Install `@woocommerce/e2e-utils` to the project as a dev dependency. This is so we can use the `shopper.addToCart` flow.
- Add `visitPostOfType` function. This will cause the e2e browser to visit a page in the WordPress dashboard by name and type. This is needed so we can navigate to a `product` page by name. It could also be useful for other tests in future.
- Add `getBlockPagePermalink` function - this will cause the e2e browser to return the permalink of the page it is on if it the page is using the block editor. The analogue to this function is `getNormalPagePermalink` which will do the same but on a page that doesn't use the block editor (for example a product page in WooCommerce).
- Add tests for the cart on the front-end.

<!-- Reference any related issues or PRs here -->
Fixes #3827

### How to test the changes in this Pull Request:

1. Add items to your cart then go to the cart block.
2. Change the quantity of an item, you may try changing multiple items, with various different quantities. **Make a note of the values!**
3. Press the "Proceed to checkout" button.
4. Use the browser's back button to get back to your cart.
5. Ensure the quantities displayed in the cart match what you expect them to from step 2.
6. Run the e2e tests and ensure all suites still pass.

<!-- If you can, add the appropriate labels -->

### Changelog

> Ensure the cart stays up to date if the user navigates using the back/forward buttons in the browser.

opr committed:
Add cartUpdate middleware

opr committed:
Include timestamp for when cart data was generated

opr committed:
Add getCartFromApi action

opr committed:
Check whether cart can be hydrated or needs to be fetched from API

opr committed:
Add cartDataIsStale action and remove getCartFromApi action

opr committed:
Add isCartDataStale selector

opr committed:
Don't load cart until staleness check is complete

opr committed:
Add comment to ease worry about locaStorage execution

opr committed:
Correct doc block and fix typographical error

opr committed:
Cater for lastCartUpdate or the timestamp being undefined

opr committed:
Include @woocommerce/api and @woocommerce/e2e-utils

opr committed:
Add getNormalPagePermalink test utility

This will allow us to get the permalink of a "normal" page, i.e. one that is not using the block editor.

opr committed:
Add getBlockPagePermalink test utility

This will get the permalink for a page using the Block editor.

opr committed:
Emit action to update cart staleness in all execution paths

opr committed:
Add visitPostOfType function

This will allow us to visit the editor page for a post based on its name and post type.

opr committed:
Add functions to get permalinks from editor pages

opr committed:
Add front-end tests for cart

opr committed:
Update package-lock.json

opr committed:
Create local function to ensure the page permalink is visible

opr committed:
Change functions for getting permalinks
They were often failing for unknown reasons, these _somehow_ make them more stable.

opr committed:
Add logging for GitHub actions

opr committed:
Add more logging for tests

opr committed:
Add more logging for tests

opr committed:
Add more logging for tests

opr committed:
Wait for networkidle on back

opr committed:
Remove logging except timestamp

opr committed:
Wait for timestamp to be added to localStorage

opr committed:
Split tests to make runs slightly shorter

opr committed:
Wait for add to cart request before continuing test

opr committed:
Fix formatting in cart.test.js

ralucaStan commented on assets/js/data/cart/action-types.ts:
```diff
@@ -4,6 +4,7 @@ export const ACTION_TYPES = {
 	REPLACE_ERRORS: 'REPLACE_ERRORS',
 	APPLYING_COUPON: 'APPLYING_COUPON',
 	REMOVING_COUPON: 'REMOVING_COUPON',
+	CART_DATA_STALE: 'CART_DATA_STALE',
```
I see that predominantly the actions start with a verb, so maybe this could be renamed to SET_CART_DATA_STALE_STATE.

opr replied:
I disagree with adding `_STATE` to the end, and I decided that `SET_IS_CART_DATA_STALE` reads better. Do you agree?

ralucaStan replied:
👍

ralucaStan commented on assets/js/data/cart/reducers.ts:
```diff
@@ -149,6 +149,16 @@ const reducer: Reducer< CartState > = (
+				...state,
+				metaData: {
+					...state.metaData,
+					cartDataIsStale: !! action.isCartDataStale,
```
``cartDataIsStale: action.isCartDataStale,`` works as well as isCartDataStale will always be Boolean

ralucaStan commented on assets/js/data/cart/selectors.ts:
```diff
@@ -59,6 +59,16 @@ export const isApplyingCoupon = ( state: CartState ): boolean => {
 };
 
+/**
+ * Returns true if cart is stale, false if it is not, and undefined if the check is not complete.
```
This will never return undefined so I guess the comment is outdated.
There is an issue tho' with not having a default value for `cartDataIsStale` on metaData as this `state.metaData.cartDataIsStale` will throw an error. 

I also see a TS warning 
<img width="802" alt="Screenshot 2021-03-23 at 16 11 24" src="https://user-images.githubusercontent.com/1628454/112179489-a619fe80-8bf2-11eb-802b-92d33ff3ac36.png">

opr replied:
This should be fixed now!

ralucaStan commented on assets/js/hocs/with-store-cart-api-hydration.js:
```diff
@@ -13,13 +13,50 @@ import { useSelect } from '@wordpress/data';
+					? null
+					: JSON.parse( lastCartUpdateRaw );
+
+			if ( lastCartUpdate?.timestamp === undefined ) {
```
If you have only one value you care about here you could directly store the timestamp and avoid working with objects. 
`lastCartUpdateTimestamp: X`

I also recommend to rename the key to `wc-blocks_cart_update_timestamp` to follow the naming pattern used for `wc-blocks_dismissed_compatibility_notices` that easily identifies the items coming from blocks.

opr replied:
Thanks, I based it on the local storage key used to track the store API nonce update which is called `storeApiNonce` - do you think it's worth adding a cooldown issue to rename this to `wc-blocks_store_api_nonce`?

ralucaStan replied:
sure, sounds good

ralucaStan commented on assets/js/hocs/with-store-cart-api-hydration.js:
```diff
@@ -13,13 +13,50 @@ import { useSelect } from '@wordpress/data';
+				cartData.current?.generated_timestamp;
+
+			if ( needsUpdateFromAPI ) {
+				cartDataIsStale();
```
Seeing this reinforces my comment to rename the action and call here something like `setCartDataStaleState(true)` as I think it brings more clarity to the code.

opr replied:
I've tidied the logic up here, please re-review and see if it makes more sense!

ralucaStan commented on assets/js/hocs/with-store-cart-api-hydration.js:
```diff
@@ -13,13 +13,50 @@ import { useSelect } from '@wordpress/data';
+				cartDataIsStale();
+				return;
+			}
+			cartDataIsStale( false );
```
I guess this line is needed because there is no default value set under metadata for cartDataIsStale. Would it be an issue to start with cartData.metaData.cartDataIsStale: false?

ralucaStan commented on assets/js/middleware/cart-update.ts:
```diff
@@ -0,0 +1,42 @@
+	return next( options, next );
+};
+
+apiFetch.use( cartUpdateMiddleware );
```
👍

ralucaStan reviewed (commented):
Thank you so much for the overview of the changes, they were really helpful for testing the PR. 👍 for the extensive work on testing.

I added some comments, let me know what you think.

opr committed:
Rename cartDataStale to setIsCartDataStale

opr committed:
Create constant for localStorage timestamp name

opr committed:
Rename cartDataIsStale to isCartDataStale

This is for consistency, the action to change this is called setIsCartDataStale - I think this reads better.

opr committed:
Use cleaner logic when determining if the cart should render or fetch

opr committed:
Change docblock for isCartDataStale selector

opr committed:
Remove boolean cast from SET_IS_CART_DATA_STALE reducer

ralucaStan commented on assets/js/hocs/with-store-cart-api-hydration.js:
```diff
@@ -13,13 +18,45 @@ import { useSelect } from '@wordpress/data';
+			if ( needsUpdateFromAPI ) {
+				setIsCartDataStale();
+			}
+		}
```
What works better here is to don't do `|| ''` for lastCartUpdateRaw. Test if there is a value and then do all the logic. This way we don't do unnecessary operations. 
```js
const lastCartUpdateRaw = window.localStorage.getItem( LAST_CART_UPDATE_TIMESTAMP_KEY );
        
if (lastCartUpdateRaw) {
        const lastCartUpdate = parseFloat( lastCartUpdateRaw );
		const cartGeneratedTimestamp = parseFloat(
		    cartData.current.generated_timestamp
	    );
        
	const needsUpdateFromAPI =
		! isNaN( cartGeneratedTimestamp ) &&
		! isNaN( lastCartUpdate ) &&
		lastCartUpdate > cartGeneratedTimestamp;
        
			        if ( needsUpdateFromAPI ) {
				        setIsCartDataStale();
			        }
}
			
```

ralucaStan commented on assets/js/data/cart/selectors.ts:
```diff
@@ -59,6 +59,16 @@ export const isApplyingCoupon = ( state: CartState ): boolean => {
+ * @return {boolean} True if the cart data is stale.
+ */
+export const isCartDataStale = ( state: CartState ): boolean => {
+	return !! state.metaData.isCartDataStale;
```
Don't think !! is needed here, these are booleans

ralucaStan reviewed (approved):
Looks good, I added a comment for a small improvement. Good to merge after that

opr committed:
Set longer timeouts for frontend cart tests

opr committed:
Enclose cart staleness checks in condition/prevent unnecessary execution

opr committed:
Remove unnecessary boolean cast from selector

opr committed:
Use constant for local storage key in tests

opr committed:
Use new localstorage key in tests. cant access constants in page context
